Added pkg/mei

insomniacslk opened the pull request:
Package `mei` adds a wrapper to MEI, the Intel ME Interface. See
https://www.kernel.org/doc/Documentation/misc-devices/mei/mei.txt .
This is a partial implementation that exposes an MEI client and an MKHI
client on top of it.

Signed-off-by: Andrea Barberio <insomniac@slackware.it>

insomniacslk commented:
I... don't understand what goes wrong. Latest `dep` installed (using the same wget command used in the circleCI config), same sequence of commands (dep check/status/ensure etc), no errors locally, but it fails on CI

hugelgupf commented on pkg/mei/mei.go:
```diff
@@ -0,0 +1,89 @@
+
+// see include/uapi/linux/mei.h
+var (
+	IoctlMEIConnectClient = ioctl.IOWR('H', 0x01, uintptr(len(ClientGUID{})))
```
Is there any way to exclude this new dependency?

hugelgupf replied:
*named constants, lol

insomniacslk replied:
it's a super-small dependency that exposes Linux's `IOWR`. I can reimplement it in this package as a small function if that's OK

insomniacslk replied:
done, more changes incoming

hugelgupf replied:
since it's super small, i'm actually okay with it. but we should also go replace pkg/ipmi's IOWR function and use this new dependency.

hugelgupf commented on pkg/mei/mkhiclient.go:
```diff
@@ -0,0 +1,300 @@
+	intelMEDevFn := 0
+
+	// check that CSE's current working state is normal
+	cs, err := txtsuite.PCIReadConfig32(intelBusID, int(intelMEDeviceID), intelMEDevFn, 0x40 /* PCI_ME_HFSTS1 */)
```
If you're just using txtsuite for pci, perhaps try pkg/pci locally? It may be missing some functionality, but it's already in our set of dependencies and would probably be easy to extend for what you need

insomniacslk replied:
yeah good point, I'll try pkg/pci (I didn't realize it's there)

insomniacslk replied:
done

hugelgupf commented on pkg/mei/mkhiclient.go:
```diff
@@ -0,0 +1,300 @@
+	// `working_state` is bits 1-4 and `operation_state` is bits 7-9.
+	// FIXME validate this code against the logic in this function's top
+	// comment.
+	if (cs&0xf) != 0x05 || ((cs>>6)&0x7) != 0x0 {
```
Can you make these constants into constants?

insomniacslk replied:
Sure.

Additionally I am doing bit-shifting here to avoid defining one `struct` with bit-field accessors for each check, but if you think it's a cleaner approach I can do that too

hugelgupf replied:
I'd probably define the struct with booleans. Read it in with marshal/unmarshal methods. Probably?

insomniacslk replied:
Any obvious downside in doing it like I did for the other `struct`s in this package? I.e. defining `type MyObject [4]byte` and having getters and setters. A bit verbose maybe, but it's a lot easier when passing data around to syscalls, or in the crazy case of I/O with unions as arguments (see ClientGUID/ClientProperties, one is for input, the other for output, same bunch of bytes)

insomniacslk replied:
defined as constants - I haven't created the structured types, I checked again and they have many fields and I'd rather not do that

hugelgupf commented on pkg/mei/mkhiclient.go:
```diff
@@ -0,0 +1,300 @@
+// GetMeiPciDevice will return the MEI PCI device object after scanning the PCI
+// bus.
+func GetMeiPciDevice() (*ghw.PCIDevice, error) {
+	pci, err := ghw.PCI()
```
pkg/pci, to avoid that long list of new dependencies

insomniacslk replied:
will do

insomniacslk replied:
done

insomniacslk commented:
> I don't know what your dep is doing, but you are vendoring u-root inside u-root...
> 
> Do you have any uncommitted directories with commands sitting around? dep doesn't care about git state - it will find dependencies for uncommitted state and gitignored state.

no other directories sitting around. Not sure why dep is vendoring u-root itself.. Let me see if anything in my .gitignore is messing up

insomniacslk commented:
Since I'm running dep from my fork (insomniacslk/u-root), maybe that's why dep also imports u-root?

insomniacslk commented:
OK - playing with symlinks in GOPATH and removing the stuff unintentionally added by the earlier runs of `dep ensure` seems to get rid of that vendored directory

insomniacslk commented:
Much better now :) it fails on some txt-suite-related stuff, but I'll get rid of that dependency as you suggested

hugelgupf commented:
Yeah, for Go projects, never clone the fork. Clone the original repo, and add your fork as a remote.

Otherwise, your Go imports are most fucked up, and will stay that way.

insomniacslk commented:
oww. It complains about %w in error format strings, that were introduced in 1.13 for error wrapping, but in u-root we support Go 1.12. I'll revert that :(

insomniacslk commented:
ready to review, PTAL

hugelgupf commented on pkg/mei/mei.go:
```diff
@@ -0,0 +1,91 @@
+
+// Open opens the specified MEI device, using the client type defined by the
+// given name. See `HECIGuids` in this package.
+func (m *MEI) Open(meiPath string, guid ClientGUID) (*ClientProperties, error) {
```
I'm not sure I fully understand the MEI API here. Could you do something like OpenMEI(...) (*MEI, *ClientProperties, error)? Or fold ClientProperties into MEI?

then you don't have the race with fd either.

hugelgupf replied:
yeah, reading further it seems to me like you could fold it into MEI, esp since it's associated specifically with that fd?

insomniacslk replied:
the client properties are client-specific (e.g. the MKHI client, basically the MEI interface lets you use different client types based on what GUID is passed via ioctl when opening the fd). I could fold it into MEI, but being MKHI-specific I'd rather leave it here. Alternatively I can move it to MEI, and access it via `m.MEI.clientProperties.SomeMethod()` instead of `m.clientProperties.SomeMethod()`

hugelgupf replied:
But you derive in MEI.Open, not in mkhi.open? That led me to believe they're MEI specific.

insomniacslk replied:
I double-checked the implementation, and they are MEI-specific, not MKHI-specific, sorry about the confusion. Yeah, I'll fold them into `MEI`

insomniacslk replied:
done in the latest push

hugelgupf commented on pkg/mei/mkhiclient.go:
```diff
@@ -0,0 +1,309 @@
+}
+
+// IsHmrfpoEnableAllowed queries whether the HMRFPO enable is allowed.
+func (m MKHIClient) IsHmrfpoEnableAllowed() (bool, error) {
```
capitalize: HMRFPO

hugelgupf replied:
give the client a pointer, *MKHIClient, or you'll be passing lots of data around with a call to this

hugelgupf commented on pkg/mei/mkhiclient.go:
```diff
@@ -0,0 +1,309 @@
+// HmrfpoEnable enables the HMRFPO (Host ME Region Flash Protection Override) via CSE,
+// see cse_hmrfpo_enable at
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L64
+func (m MKHIClient) HmrfpoEnable() error {
```
EnableHMRFPO

hugelgupf replied:
pointer

hugelgupf commented on pkg/mei/mkhihdr.go:
```diff
@@ -0,0 +1,63 @@
+
+package mei
+
+// MKHIHdr is the header for every MKHI command, see
```
do you expect commands for the ME to be implemented outside this package? if not, i'd unexport this

insomniacslk replied:
I will unexport

hugelgupf commented on pkg/mei/mkhihdr.go:
```diff
@@ -0,0 +1,63 @@
+// Copyright 2012-2017 the u-root Authors. All rights reserved
```
2020

hugelgupf commented on pkg/mei/mkhiclient.go:
```diff
@@ -0,0 +1,309 @@
+
+// GetMeiPciDevice will return the MEI PCI device object after scanning the PCI
+// bus.
+func GetMeiPciDevice() (*pci.PCI, error) {
```
does this need to be exported?

insomniacslk replied:
not really, it's nice to have though, in case other packages need to wrap on top of MEI. I don't have strong opinions on this

hugelgupf commented on pkg/mei/mkhiclient.go:
```diff
@@ -0,0 +1,309 @@
+	log.Printf("MEI Device found: %s", meiDev)
+
+	// check that CSE's current working state is normal
+	cs, err := meiDev.ReadConfigRegister(0x40 /* PCI_ME_HFSTS1*/, 32)
```
define 0x40 and 0x60 at least as named constants in this file

insomniacslk replied:
whoops, missed that, thanks

hugelgupf commented on pkg/mei/mkhiclient.go:
```diff
@@ -0,0 +1,309 @@
+	hdr.SetCommand(MKHIHmrfpoEnable)
+	canEnable, err := m.IsHmrfpoEnableAllowed()
+	if err != nil {
+		return fmt.Errorf("Enabling HMRFPO failed: %v", err)
```
don't cap error strings: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

insomniacslk replied:
good catch, it was a `log.Printf` in a previous life

hugelgupf commented on pkg/mei/mkhiclient.go:
```diff
@@ -0,0 +1,309 @@
+}
+
+// Open opens an MKHI client connection.
+func (m *MKHIClient) Open(meiPath string) error {
```
why not 

OpenMKHIClient(meiPath string) (*MKHIClient, error)

?

insomniacslk replied:
I would rather use top-level functions for `New*` (not needed here, it's all zero-values) instead than for `Open`, since `Open` could be called at a later stage

hugelgupf replied:
I don't understand, what later stage? What could be called before Open?

hugelgupf replied:
You can't call any of the other methods without Open, as far as I can tell.

insomniacslk replied:
I'm thinking more of scenarios of embedding this struct, where `MKHIClient` is a struct field and Open is called at a later stage with `s.mkhi.Open()` rather than `mei.OpenMKHIClient(s.mkhi)` which feels more natural to use

hugelgupf replied:
I'm thinking it should be 

mkhi, err := OpenMKHI

s.mkhi = mkhi

As well as

mei, err := OpenMEI

I see this kind of as a poor man's RAII - struct doesn't exist when not initialized, so it cannot be misused when uninitialized

hugelgupf commented on pkg/mei/mkhiclient.go:
```diff
@@ -0,0 +1,309 @@
+// MKHI command groups, see
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L22
+const (
+	MKHIGroupIDCbm    = 0x0
```
do these need to be exported?

insomniacslk replied:
not really, I'll unexport

hugelgupf commented on pkg/mei/mkhiclient.go:
```diff
@@ -0,0 +1,309 @@
+// various ME-related constants
+const (
+	// ME - current working state set to normal
+	MEHfs1CwsNormal = 0x5
```
do these need to be exported?

insomniacslk replied:
no, will unexport

hugelgupf commented on pkg/mei/mkhiclient.go:
```diff
@@ -0,0 +1,309 @@
+	MEHfs3FwSkuCustom = 0x5
+)
+
+// MKHIClient is a client to send MKHI commands via MEI.
```
what, in fact, are MKHI commands?

insomniacslk replied:
I wish I knew the acronym :( all of these come from reverse engineering apparently

hugelgupf replied:
Hm I somehow doubt that but nevermind I looked it up.

zaolin commented on pkg/mei/ioctl.go:
```diff
@@ -0,0 +1,77 @@
+// Copyright 2012-2017 the u-root Authors. All rights reserved
```
2020

zaolin commented on pkg/mei/mei.go:
```diff
@@ -0,0 +1,91 @@
+// Copyright 2012-2017 the u-root Authors. All rights reserved
```
same here

hugelgupf commented on pkg/mei/mei.go:
```diff
@@ -0,0 +1,91 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mei
```
Write a package comment. I'd mention the device name /dev/mei0, and what kernel configs to turn on. We haven't been doing that and it'd be nice to start...

insomniacslk replied:
done! I think all these files should also have a `_linux.go` extension since this is linux-specific

insomniacslk commented:
Addressed all the feedback, except:
* folding ClientProperties into MEI, since it's MKHI-specific, https://github.com/u-root/u-root/pull/1642#discussion_r411034438
* GetMeiPciDevice left public, https://github.com/u-root/u-root/pull/1642#discussion_r411035944
* keep using Open instead of top-level function, https://github.com/u-root/u-root/pull/1642#discussion_r411036457

Also, this is all Linux-specific, should all the files end with `_linux.go` ?

hugelgupf commented:
yeah, they should probably all end in _linux.go

insomniacslk commented:
Folded `ClientProperties` in MEI instead of MKHIClient, and changed files to end in `_linux.go`

pmazzini commented on pkg/mei/ioctl_linux.go:
```diff
@@ -0,0 +1,77 @@
+// this file contains a few ioctl-related constants and reimplementations of
+// macros from Linux's include/uapi/asm-generic/ioctl.h .
+
+const (
```
Maybe we should have a separate ioctl package to avoid duplicating this:
https://github.com/u-root/u-root/blob/master/pkg/ipmi/ioctl_unix.go

insomniacslk replied:
yes, that's what @hugelgupf suggested. I can fall back to the original `ioctl` package as discussed in https://github.com/u-root/u-root/pull/1642#discussion_r411036949 , and in another PR migrate `ipmi` to it

insomniacslk replied:
done

insomniacslk commented:
I also need to swap `File.Write` with `syscall.Write` to avoid `epoll` as found by @zaolin

insomniacslk commented:
This is ready to review again

hugelgupf commented:
I can't seem to unresolve this part, so I'm recommending

I'm thinking it should be

mkhi, err := OpenMKHI

s.mkhi = mkhi

As well as

mei, err := OpenMEI

I see this kind of as a poor man's golang RAII - struct doesn't exist when not initialized, so it cannot be misused or racy when uninitialized

hugelgupf commented on pkg/mei/mei_linux.go:
```diff
@@ -0,0 +1,99 @@
+// Open opens the specified MEI device, using the client type defined by the
+// given name. See `HECIGuids` in this package.
+func (m *MEI) Open(meiPath string, guid ClientGUID) error {
+	fd, err := os.OpenFile(meiPath, os.O_RDWR, 0755)
```
os.OpenFile will still set the FD to be non-blocking and add it to the epoll, despite using the sys Write and Read below. You want it to be a blocking fd, I assume.

To do that, do a syscall.Open and use os.NewFile to make a *os.File out of it. Then you should be even able to use File.Read/Write again.

insomniacslk replied:
good point - we tested the code and it actually does what we expect, but uniforming on the same API is better

insomniacslk replied:
done

insomniacslk commented:
I'll implement OpenMEI and OpenMKHI. Thanks for your review!

insomniacslk commented:
Switched to OpenMEI/OpenMKHI. PTAL @hugelgupf

hugelgupf commented on pkg/mei/mkhiclient_linux.go:
```diff
@@ -0,0 +1,309 @@
+
+// pciAddressToInt converts the last part of a PCI address string to integer,
+// e.g. 0000:00:16.0 will return 0x16 .
+func pciAddressToInt(pciaddr string) (int, error) {
```
I'm so sorry, one more... I don't see this func used anywhere, can it be deleted?

insomniacslk replied:
whoops, sorry! It was used in a previous version of this code, just removed and committed again

insomniacslk commented:
Thank you so much @hugelgupf for the review!
